restapi/src/main/java/io/stargate/web/docsapi/service/JsonShreddedRow.java

```diff
@@ -0,0 +1,63 @@
+/*
+ * Copyright The Stargate Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package io.stargate.web.docsapi.service;
+
+import io.stargate.web.docsapi.exception.ErrorCode;
+import io.stargate.web.docsapi.exception.ErrorCodeRuntimeException;
+import java.util.List;
+import javax.annotation.Nullable;
+import org.immutables.value.Value;
+
+@Value.Immutable
+public interface JsonShreddedRow {
+
+  // TODO Convert to abstract class, hide?
+
+  @Value.Check
+  default void validate() {
+    if (getPath().size() > getMaxDepth()) {
+      throw new ErrorCodeRuntimeException(ErrorCode.DOCS_API_GENERAL_DEPTH_EXCEEDED);
+    }
+  }
+
+  String getKey();
+
+  int getMaxDepth();
+
+  List<String> getPath();
+
+  // leaf is always actually the last path we have
+  default String getLeaf() {
+    List<String> path = getPath();
+    if (path != null && !path.isEmpty()) {
+      return path.get(path.size() - 1);
+    } else {
+      throw new IllegalStateException("Shredded row does not contain a single path.");
+    }
+  }
+
+  @Nullable
+  String getStringValue();
```

Comment:
Since we're already boxing/unboxing these values. What do you think about using a "tagged union" representation?

```java
@Value.Immutable
public interface JsonShreddedRow {

  enum ValueType {
    Unknown,
    String,
    Double,
    Boolean,
  }

  ValueType type();

  Object value();
}
```

The representation is a bit more compact in the generated immutable code and the value extraction comparisons might be a bit faster (maybe).

```
    switch (row.type()) {
      case Unknown:
        // Other stuff
      case String:
        // String stuff
      case Double:
        // Double stuff
      case Boolean:
        // Boolean stuff
    }

    if (row.getStringValue() != null) {
      // String stuff
    } else if (row.getDoubleValue() != null) {
      // Double stuff
    } else if (row.getBooleanValue() != null) {
      // Boolean stuff
    } else {
      // Other stuff
    }
```

Reply:
Nope, I need all the values when storing the row as we do have three columns for the values.. So for me the switch does not make any sense.

Reply:
Ah, makes sense. So we store more than one of these values in a row?

Reply:
No it's only one.. But I still need to bind the others to `null`..

Reply:
Why do we need to bind the others to `null`? Can't they be left unset?

Reply:
Just for the record, because the type of the JSON field can change on update, so we need to count for that upfront.